### PR TITLE
[IOTDB-141]Develop memory estimate tool

### PR DESCRIPTION
**Pull request by fanhualta**
Currently, we have a [dynamic parameter adapter module](https://github.com/apache/incubator-iotdb/pull/232), so based on this module, it's better to develop a memory estimation tool for users to use. Users provide the number of storage groups and the number of timeseries according to their workloads, and the tool gives the minimum memory for writing to satisfy the load.
jira link: https://issues.apache.org/jira/browse/IOTDB-141

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -126,6 +126,11 @@
             <artifactId>jaxb-runtime</artifactId>
             <version>2.4.0-b180725.0644</version>
         </dependency>
+        <dependency>
+            <groupId>io.airlift</groupId>
+            <artifactId>airline</artifactId>
+            <version>0.8</version>
+        </dependency>
```
Comment (qiaojialin) [inline pom.xml]: please check the license of this dependency, and update the LICENSE and NOTICE file

Comment (fanhualta) [reply pom.xml]: Actually, it has ASF license.

**Comment by fanhualta** [comment]
> We need an example or hint when using this tool, you can add it in the documents

I have written a docu about this tool.

**Comment by samperson1997** [comment]
Documents in User Guide and file `0-content.md` is also changed in [PR#247](https://github.com/apache/incubator-iotdb/pull/247). Maybe the order of new chapters (Chapter 7 and Chapter 8) needs to be discussed first to avoid conflict.

**Comment by fanhualta** [comment]
> Documents in User Guide and file `0-content.md` are also changed in [PR#247](https://github.com/apache/incubator-iotdb/pull/247). Maybe the order of new chapters (Chapter 7 and Chapter 8) needs to be discussed first to avoid conflict.

I think 7 is for all tolls.

File: pom.xml (hunk 1 above)
Comment (jixuan1989) [reply pom.xml]: "Airlift is a framework for building REST services in Java." 
Ah? did you add a REST service??
by the way, we need to add comments when we add new dependencies.

Comment (fanhualta) [reply pom.xml]: No， I only use its airline module.

File: docs/Documentation-CHN/UserGuideV0.7.0/7-System Tools/1-Sync.md [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+   </tr>
+</table>
+
+<table>
+   <tr>
+      <td colspan="2">参数名: sync_server_port</td>
+   </tr>
+   <tr>
+      <td width="20%">描述</td>
+      <td>同步接收端服务器监听接口，请确认该端口不是系统保留端口并且未被占用。参数is_postback_enable设置为true时有效，参数is_postback_enable设置为false时无效</td>
```
[the remark this answers is not in the document]
Comment (fanhualta) [reply docs/Documentation-CHN/UserGuideV0.7.0/7-System Tools/1-Sync.md]: sync, fixed

File: docs/Documentation-CHN/UserGuideV0.7.0/7-System Tools/1-Sync.md [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+   </tr>
+   <tr>
+      <td>改后生效方式</td>
+      <td>重启同步功能发送端生效</td>
+   </tr>
+</table>
+
+<table>
+   <tr>
+      <td colspan="2">参数名: iotdb_bufferWrite_directory</td>
```
[the remark this answers is not in the document]
Comment (fanhualta) [reply docs/Documentation-CHN/UserGuideV0.7.0/7-System Tools/1-Sync.md]: This issue will be modified in sync pr, not currently.

File: docs/Documentation-CHN/UserGuideV0.7.0/7-System Tools/2-Memory Estimation Tool.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+在内存预估时，若工具计算需要较长的时间，则会在下方显示出运行进度，便于用户掌握进度。
+
+# 使用方式
+
+用户可以使用```$IOTDB_HOME/bin```文件夹下的脚本使用该工具
+Linux系统与MacOS系统启动命令如下：
+* 以20个存储组，共10w条时间序列，时间序列在存储组中均分为例：
+```
+  Shell >$IOTDB_HOME/bin/memory-tool.sh -sg 20 -ts 100000 calmem
```
[the remark this answers is not in the document]
Comment (fanhualta) [reply docs/Documentation-CHN/UserGuideV0.7.0/7-System Tools/2-Memory Estimation Tool.md]: help

File: docs/Documentation/UserGuideV0.7.0/7-System Tools/1-Sync.md [hunk 5]
```diff
@@ -54,7 +58,7 @@ The parameter configuration of the sync receiver is located in the configuration
    </tr>
    <tr>
       <td width="30%">Description</td>
-      <td>Sync function switch, which is configured as true to indicate that the receiver is allowed to receive the data from sender and load it. When set to false, it means that the receiver is not allowed to receive the data from any sender.</td>
+      <td>Sync function switch, which is configured as true to indicate that the receiver is allowed to receive the data from sender and load it. When set to false, it means that the receiver is not allowed to receive the data from any sender. </td>
```
[the remark this answers is not in the document]
Comment (fanhualta) [reply docs/Documentation/UserGuideV0.7.0/7-System Tools/1-Sync.md]: fixed

File: iotdb/src/main/java/org/apache/iotdb/db/conf/adapter/IoTDBConfigDynamicAdapter.java [hunk 6]
```diff
@@ -37,10 +36,11 @@
  *
  * 1. maxMemTableNum. This parameter represents the size of the MemTable available in the MemTable
```
[the remark this answers is not in the document]
Comment (fanhualta) [reply iotdb/src/main/java/org/apache/iotdb/db/conf/adapter/IoTDBConfigDynamicAdapter.java]: Actually, it's the name of the parameter, its concrete meaning is explained in the following sentence.

File: docs/Documentation-CHN/UserGuideV0.7.0/7-System Tools/2-Memory Estimation Tool.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (jt2594838) [reply docs/Documentation-CHN/UserGuideV0.7.0/7-System Tools/2-Memory Estimation Tool.md]: You'd better point it out in the document.

File: pom.xml (hunk 1 above)
Comment (liuruiyiyang) [reply pom.xml]: Is this library's function similar to 
```
<dependency>
           <groupId>commons-cli</groupId>
            <artifactId>commons-cli</artifactId>
            <version>${common.cli.version}</version>
 </dependency>
```
in iotdb-cli module ?
Should we choose one and keep our dependencies as less as possible?

Comment (fanhualta) [reply pom.xml]: I have seen the implementation in cli module, and I think this package is more useful and extensible; one for IoTDB module, one for Cli module, it should be fine.

File: iotdb/src/main/java/org/apache/iotdb/db/conf/adapter/IoTDBConfigDynamicAdapter.java [hunk 7]
```diff
@@ -135,8 +129,13 @@ public synchronized boolean tryToAdaptParameters() {
...
       tsFileSize = calcTsFileSize(memTableSizeFloorThreshold);
-      memtableSizeInByte = memTableSizeFloorThreshold + ((tsFileSize - memTableSizeFloorThreshold) >> 1);
-      if (tsFileSize < memTableSizeFloorThreshold) {
+      if(tsFileSize > memTableSizeFloorThreshold) {
+        memtableSizeInByte =
+            memTableSizeFloorThreshold + ((tsFileSize - memTableSizeFloorThreshold) >> 1);
+      } else {
+        memtableSizeInByte = memTableSizeFloorThreshold;
+      }
+      if (tsFileSize * CompressionRatio.getInstance().getRatio() < memTableSizeFloorThreshold) {
```
[the remark this answers is not in the document]
Comment (fanhualta) [reply iotdb/src/main/java/org/apache/iotdb/db/conf/adapter/IoTDBConfigDynamicAdapter.java]: I've modified these codes. You read it again to see if you can understand it.

**Pull request merged**